// @wix/wix-custom-elements/dist/esmTarget/elements/wixElement.js
function _createForOfIteratorHelperLoose(o, allowArrayLike) {
    var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"];
    if (it) return (it = it.call(o)).next.bind(it);
    if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") {
        if (it) o = it;
        var i = 0;
        return function() {
            if (i >= o.length) return {
                done: true
            };
            return {
                done: false,
                value: o[i++]
            };
        };
    }
    throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}

function _unsupportedIterableToArray(o, minLen) {
    if (!o) return;
    if (typeof o === "string") return _arrayLikeToArray(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor) n = o.constructor.name;
    if (n === "Map" || n === "Set") return Array.from(o);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
    if (len == null || len > arr.length) len = arr.length;
    for (var i = 0, arr2 = new Array(len); i < len; i++) {
        arr2[i] = arr[i];
    }
    return arr2;
}

function _inheritsLoose(subClass, superClass) {
    subClass.prototype = Object.create(superClass.prototype);
    subClass.prototype.constructor = subClass;
    _setPrototypeOf(subClass, superClass);
}

function _wrapNativeSuper(Class) {
    var _cache = typeof Map === "function" ? new Map() : undefined;
    _wrapNativeSuper = function _wrapNativeSuper(Class) {
        if (Class === null || !_isNativeFunction(Class)) return Class;
        if (typeof Class !== "function") {
            throw new TypeError("Super expression must either be null or a function");
        }
        if (typeof _cache !== "undefined") {
            if (_cache.has(Class)) return _cache.get(Class);
            _cache.set(Class, Wrapper);
        }

        function Wrapper() {
            return _construct(Class, arguments, _getPrototypeOf(this).constructor);
        }
        Wrapper.prototype = Object.create(Class.prototype, {
            constructor: {
                value: Wrapper,
                enumerable: false,
                writable: true,
                configurable: true
            }
        });
        return _setPrototypeOf(Wrapper, Class);
    };
    return _wrapNativeSuper(Class);
}

function _construct(Parent, args, Class) {
    if (_isNativeReflectConstruct()) {
        _construct = Reflect.construct;
    } else {
        _construct = function _construct(Parent, args, Class) {
            var a = [null];
            a.push.apply(a, args);
            var Constructor = Function.bind.apply(Parent, a);
            var instance = new Constructor();
            if (Class) _setPrototypeOf(instance, Class.prototype);
            return instance;
        };
    }
    return _construct.apply(null, arguments);
}

function _isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;
    try {
        Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {}));
        return true;
    } catch (e) {
        return false;
    }
}

function _isNativeFunction(fn) {
    return Function.toString.call(fn).indexOf("[native code]") !== -1;
}

function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
        o.__proto__ = p;
        return o;
    };
    return _setPrototypeOf(o, p);
}

function _getPrototypeOf(o) {
    _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
        return o.__proto__ || Object.getPrototypeOf(o);
    };
    return _getPrototypeOf(o);
}

function wixElementWrapper(services) {
    var WixElement = /*#__PURE__*/ function(_HTMLElement) {
        _inheritsLoose(WixElement, _HTMLElement);

        function WixElement() {
            // eslint-disable-line no-useless-constructor
            return _HTMLElement.call(this) || this;
        }

        var _proto = WixElement.prototype;

        _proto.reLayout = function reLayout() { //should be implemented inside child element
        };

        _proto.connectedCallback = function connectedCallback() {
            this.observeResize();
            this.reLayout();
        };

        _proto.disconnectedCallback = function disconnectedCallback() {
            this.unobserveResize();
            this.unobserveChildren();
        };

        _proto.observeResize = function observeResize() {
            services.resizeService.observe(this);
        };

        _proto.unobserveResize = function unobserveResize() {
            services.resizeService.unobserve(this);
        }
        /**01
         * Observe DOM mutations to wait for addition of missing children
         *
         * @param {HTMLElement} el
         */
        ;

        _proto.observeChildren = function observeChildren(el) {
            var _this = this;

            if (!this.childListObserver) {
                this.childListObserver = new MutationObserver(function() {
                    return _this.reLayout();
                });
            }

            this.childListObserver.observe(el, {
                childList: true
            });
        };

        _proto.observeChildAttributes = function observeChildAttributes(el, attributes) {
            var _this2 = this;

            if (attributes === void 0) {
                attributes = [];
            }

            if (!this.childrenAttributesObservers) {
                this.childrenAttributesObservers = [];
            }

            var attributesObserver = new MutationObserver(function() {
                return _this2.reLayout();
            });
            attributesObserver.observe(el, {
                attributeFilter: attributes
            });
            this.childrenAttributesObservers.push(attributesObserver);
        };

        _proto.observeChildResize = function observeChildResize(child) {
            if (!this.childrenResizeObservers) {
                this.childrenResizeObservers = [];
            }

            services.resizeService.observeChild(child, this);
            this.childrenResizeObservers.push(child);
        };

        _proto.unobserveChildrenResize = function unobserveChildrenResize() {
            if (this.childrenResizeObservers) {
                this.childrenResizeObservers.forEach(function(child) {
                    services.resizeService.unobserveChild(child);
                });
                this.childrenResizeObservers = null;
            }
        }
        /**
         * Remove DOM MutationObserver if one was created
         */
        ;

        _proto.unobserveChildren = function unobserveChildren() {
            if (this.childListObserver) {
                this.childListObserver.disconnect();
                this.childListObserver = null;
            }

            if (this.childrenAttributesObservers) {
                for (var _iterator = _createForOfIteratorHelperLoose(this.childrenAttributesObservers), _step; !(_step = _iterator()).done;) {
                    var observer = _step.value;
                    observer.disconnect();
                    observer = null;
                }

                this.childrenAttributesObservers = null;
            }

            this.unobserveChildrenResize();
        };

        return WixElement;
    }( /*#__PURE__*/ _wrapNativeSuper(HTMLElement));

    return WixElement;
}

export default wixElementWrapper;
