// @wix/santa-animations/src/definitions/animation/in/mobile/cornerIn.js
'use strict'
const {
    setSmoothTransformOrigin
} = require('../../../../utils/definitionsUtils')

const name = 'CornerIn'
const properties = {
    hideOnStart: true,
    mobile: true,
    viewportThreshold: 0.15,

    groups: ['entrance', 'animation'],
    schema: {
        duration: {
            type: 'number',
            min: 0,
            default: 0
        },
        delay: {
            type: 'number',
            min: 0,
            default: 0
        },
        direction: {
            type: 'string',
            enum: ['right', 'left'],
            default: 'right'
        }
    }
}

const paramsMap = {
    left: {
        dx: '-1',
        angle: '45'
    },
    right: {
        dx: '1',
        angle: '45'
    }
}

function register({
    factory
}) {
    /**
     * TurnIn animation object
     * @param {Array<HTMLElement>|HTMLElement} elements DOM element to animate
     * @param {Number} [duration]
     * @param {Number} [delay]
     * @param {Object} [params]
     * @param {'left'|'right'} [direction='right'] 'left' or 'right'
     * @returns {TimelineMax}
     */
    function animation(elements, duration, delay, {
        direction = properties.schema.direction.default,
        ...params
    } = {}) {
        const origin = paramsMap[direction]
        const transformRotate = (origin.dx > 0 ? '+=' : '-=') + origin.angle

        const sequence = factory.sequence(params)
        sequence.add(factory.animate('BaseFade', elements, duration, delay, {
            from: {
                opacity: 0
            },
            to: {
                opacity: 1
            },
            ease: 'Linear.easeIn'
        }))

        elements.forEach(element => {
            const transformOrigin = origin.dx > 0 ? '100% 100%' : '0 100%'
            sequence.add([
                setSmoothTransformOrigin(element, transformOrigin, factory.animate),
                factory.animate('BaseRotate', element, duration, delay, {
                    from: {
                        rotation: transformRotate
                    },
                    ease: 'Quad.easeOut',
                    immediateRender: false
                })
            ], 0)
        })

        return sequence.get()
    }

    factory.registerAnimation(name, animation, properties)
}

module.exports = {
    name,
    properties,
    register
}
