// @wix/wix-custom-elements/dist/esmTarget/elements/wixBgImage/wixBgImage.js
function _defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}

function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}

function _inheritsLoose(subClass, superClass) {
    subClass.prototype = Object.create(superClass.prototype);
    subClass.prototype.constructor = subClass;
    _setPrototypeOf(subClass, superClass);
}

function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
        o.__proto__ = p;
        return o;
    };
    return _setPrototypeOf(o, p);
}

import imageLayout from './bgImageLayout';
import {
    shouldStopImageLoad
} from '../../utils/imageUtils';

function wixBgImageWrapper(WixElement, services, environmentConsts) {
    var WixBgImage = /*#__PURE__*/ function(_WixElement) {
        _inheritsLoose(WixBgImage, _WixElement);

        function WixBgImage() {
            // eslint-disable-line no-useless-constructor
            return _WixElement.call(this) || this;
        }

        var _proto = WixBgImage.prototype;

        _proto.reLayout = function reLayout() {
            //todo: temp experiment for measuring purposes, should be removed after test results
            if (shouldStopImageLoad(services)) {
                return;
            }

            var domNodes = {};
            var measures = {};
            var imageId = this.getAttribute('id');
            var imageInfo = JSON.parse(this.dataset.tiledImageInfo);
            var bgEffectName = this.dataset.bgEffectName;
            var containerId = imageInfo.containerId;
            var container = document.getElementById(containerId);
            domNodes[imageId] = this;
            domNodes[containerId] = container;
            imageInfo.displayMode = imageInfo.imageData.displayMode;
            services.mutationService.measure(function() {
                imageLayout.measure(imageId, measures, domNodes, {
                    containerId: containerId,
                    bgEffectName: bgEffectName
                }, services);
            });
            services.mutationService.mutate(function() {
                imageLayout.patch(imageId, measures, domNodes, imageInfo, environmentConsts);
            });
        };

        _proto.attributeChangedCallback = function attributeChangedCallback(name, oldValue) {
            if (oldValue) {
                this.reLayout();
            }
        };

        _proto.disconnectedCallback = function disconnectedCallback() {
            _WixElement.prototype.disconnectedCallback.call(this);
        };

        _createClass(WixBgImage, null, [{
            key: "observedAttributes",
            get: function get() {
                return ['data-tiled-image-info'];
            }
        }]);

        return WixBgImage;
    }(WixElement);

    return WixBgImage;
}

export default wixBgImageWrapper;
